### DISCUSSION TITLE
possibility to inject custom attributes of element types to field layout

### DISCUSSION THREAD

**Discussion by vnali** [craftcms/cms#10334] [Q&A] [answered]
I am working on a custom element type which has custom attributes.

currently to render this custom attributes i create tab headers and pass to twig and render tab contents in twig.
some disadvantage are:
- i can not  inject my custom attributes in middle of custom fields.
- i can not render UI elements.

```
<div id="tab1">

// render custom attributes 
{{ forms.textField({
	label: 'Weight',
	id: 'weight',
	name: 'weight',
	required: true,
	errors: myelement.getErrors('weight'),
	value: myelement.weight,
	translatable: true,
	siteId: myelement.siteId
}) }}

// render fields on tab1
{% include '_includes/fields' with {
	fields:  tab1Fields,
	element: myelement,
	static: false
} only %}
```

now i am thinking about a way to inject my custom attributes to field layout tabs so i can use `$fieldsHtml = $form->render();` to render everything in field layout.

i tried to put my custom attributes in separate custom.twig file and inject it in field layout via Templete element UI but it seems it only accepts project root templates -probably because of security concerns, and also i am not sure if my element is accessible this way to custom.twig file-.

is there any way to programmatically attach element type's custom attributes to field layout?

**Comment by brandonkelly** [comment, marked as answer]
You can do this by registering your attributes’ fields as “standard fields” on the field layout, as mandatory layout elements. (They will be called “native fields” in Craft 4, but otherwise works the same.)

You can do that with the [ EVENT_DEFINE_STANDARD_FIELDS](https://docs.craftcms.com/api/v3/craft-models-fieldlayout.html#event-define-standard-fields) event from your plugin’s `init()` method:

```php
use craft\events\DefineFieldLayoutFieldsEvent;
use craft\fieldlayoutelements\StandardTextField;
use craft\models\FieldLayout;
use yii\base\Event;

Event::on(
    FieldLayout::class,
    FieldLayout::EVENT_DEFINE_STANDARD_FIELDS,
    function(DefineFieldLayoutFieldsEvent $event) {
        /** @var FieldLayout $fieldLayout */
        $fieldLayout = $event->sender;

        if ($fieldLayout->type === MyElementType::class) {
            $event->fields[] = [
                'class' => StandardTextField::class,
                'attribute' => 'width',
                'mandatory' => true, // whether the field *must* be included in field layouts
                'required' => true, // whether the field should show that it is required
                'translatable' => true,
            ];
        }
    }
);
```

That example is for a text field, using [`craft\fieldlayoutelements\StandardTextField`](https://docs.craftcms.com/api/v3/craft-fieldlayoutelements-standardtextfield.html). If you have a need for a different type of field, you will need to create the field layout element class yourself, and have it extend [`craft\fieldlayoutelements\StandardField`](https://docs.craftcms.com/api/v3/craft-fieldlayoutelements-standardfield.html).

**Comment by vnali** [comment]
@brandonkelly 
it's great!
there is some things which i think this is not possible with this event but nice to have them:
- removing standard field from field layout via this event 
- specifying the tab name that we want for standard fields
- changing attributes of standard fields via this event.  //  i am not sure, it seems changing attributes works sometimes
